src/other/ThreadedRadialGreensConvolutionLoop.C

```diff
@@ -0,0 +1,169 @@
+/**********************************************************************/
+
+#include "ThreadedRadialGreensConvolutionLoop.h"
+#include "Function.h"
+
+ThreadedRadialGreensConvolutionLoop::ThreadedRadialGreensConvolutionLoop(
+    RadialGreensConvolution & green)
+  : _green(green), _function_name(_green.parameters().get<FunctionName>("function"))
+{
+}
+
+// Splitting Constructor
+ThreadedRadialGreensConvolutionLoop::ThreadedRadialGreensConvolutionLoop(
+    const ThreadedRadialGreensConvolutionLoop & x, Threads::split /*split*/)
+  : _green(x._green), _function_name(x._function_name)
+{
+}
+
+void
+ThreadedRadialGreensConvolutionLoop::operator()(const QPDataRange & qpdata_range)
+{
+  // fetch data from parent
+  const auto r_cut = _green._r_cut;
+  const auto & qp_data = _green._qp_data;
+  const auto & correction_integral = _green._correction_integral;
+  const auto & kd_tree = _green._kd_tree;
+  const auto dim = _green._dim;
+  const auto t = _green._t;
+
+  // fetch thread copy of the function for safe evaluation
+  ParallelUniqueId puid;
+  Function & function = _green._fe_problem.getFunction(_function_name, puid.id);
+
+  // radial bin size
+  const Real dr = r_cut / correction_integral.size();
+
+  // integral of the convolution (used if normalization is requested)
+  _convolution_integral = 0.0;
+
+  // tree search data structures
+  std::vector<std::pair<std::size_t, Real>> ret_matches;
+  nanoflann::SearchParams search_params;
+
+  // result map entry
+  const auto end_it = _green._convolution.end();
+  auto it = end_it;
+
+  // iterate over qp range
+  for (auto && local_qp : qpdata_range)
+  {
+    // Look up result map iterator only if we enter a new element. this saves a bunch
+    // of map lookups because same element entries are consecutive in the qp_data vector.
+    if (it == end_it || it->first != local_qp._elem_id)
+      it = _green._convolution.find(local_qp._elem_id);
+
+    // initialize result entry
+    mooseAssert(it != end_it, "Current element id not found in result set.");
+    auto & sum = it->second[local_qp._qp];
+    sum = 0.0;
+
+    // if the variable is periodic we need to perform extra searches translated onto
+    // the periodic neighbors
+    std::list<Point> cell_vector = {Point()};
+    for (unsigned int j = 0; j < dim; ++j)
+      if (_green._periodic[j])
+      {
+        std::list<Point> new_cell_vector;
+
+        for (const auto & cell : cell_vector)
+        {
+          if (local_qp._q_point(j) + _green._periodic_vector[j](j) - r_cut <
+              _green._periodic_max[j])
+            new_cell_vector.push_back(cell + _green._periodic_vector[j]);
+
+          if (local_qp._q_point(j) - _green._periodic_vector[j](j) + r_cut >
+              _green._periodic_min[j])
+            new_cell_vector.push_back(cell - _green._periodic_vector[j]);
+        }
+
+        cell_vector.insert(cell_vector.end(), new_cell_vector.begin(), new_cell_vector.end());
```

Comment:
potential optimization: `std::list::splice`

Reply:
Yes! splice turns out to be about twice as fast in this scenario here. I'll make a PR with that change. There are other places in the code where splice could help, too.